pkg/inframetadata/internal/hostmap/hostmap.go

```diff
@@ -0,0 +1,157 @@
+)
+
+// HostMap maps from hostnames to host metadata payloads.
+type HostMap struct {
+	// mu is the mutex for the host map and updater.
+	mu sync.Mutex
+	// hosts map
+	hosts map[string]payload.HostMetadata
+}
+
+// New creates a new HostMap.
+func New() *HostMap {
+	return &HostMap{
+		hosts: make(map[string]payload.HostMetadata),
+	}
+}
+
+// strField gets a string-typed field from a resource attribute map.
+// It returns:
+// - The field's value, if available
+// - Whether the field was present in the map
+// - Any errors found in the process
+func strField(m pcommon.Map, key string) (string, bool, error) {
+	val, ok := m.Get(key)
+	if !ok {
+		// Field not available, don't update but don't fail either
+		return "", false, nil
+	}
+	if val.Type() != pcommon.ValueTypeStr {
+		return "", false, fmt.Errorf("%q has type %q, expected type \"Str\" instead", key, val.Type())
+	}
+	return val.Str(), true, nil
+}
+
+// isAWS checks if a resource attribute map
+// is coming from an AWS VM.
+func isAWS(m pcommon.Map) (bool, error) {
+	cloudProvider, ok, err := strField(m, conventions.AttributeCloudProvider)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		// no cloud provider field
+		return false, nil
+	}
+	return cloudProvider == conventions.AttributeCloudProviderAWS, nil
+}
+
+// instanceID gets the AWS EC2 instance ID from a resource attribute map.
+// It returns:
+// - The EC2 instance id if available
+// - Whether the instance id was found
+// - Any errors found retrieving the ID
+func instanceID(m pcommon.Map) (string, bool, error) {
+	if onAWS, err := isAWS(m); err != nil || !onAWS {
+		return "", onAWS, err
+	}
+	return strField(m, conventions.AttributeHostID)
+}
+
+// ec2Hostname gets the AWS EC2 OS hostname from a resource attribute map.
+// It returns:
+// - The EC2 OS hostname if available
+// - Whether the EC2 OS hostname was found
+// - Any errors found retrieving the ID
+func ec2Hostname(m pcommon.Map) (string, bool, error) {
+	if onAWS, err := isAWS(m); err != nil || !onAWS {
+		return "", onAWS, err
+	}
+	return strField(m, conventions.AttributeHostName)
+}
+
+// Update the information about a given host by providing a resource.
+// The function reports:
+//   - Whether the information about the `host` has changed
+//   - Any non-fatal errors that may have occurred during the update
+//
+// Partial modifications will still be applied even with non-fatal errors.
+func (m *HostMap) Update(host string, res pcommon.Resource) (changed bool, err error) {
+	md := payload.HostMetadata{
+		Flavor:  "otelcol-contrib",
+		Meta:    &payload.Meta{},
+		Tags:    &payload.HostTags{},
+		Payload: gohai.NewEmpty(),
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var found bool
+	if old, ok := m.hosts[host]; ok {
+		found = true
+		md = old
+	}
+
+	md.InternalHostname = host
+	md.Meta.Hostname = host
+
+	// InstanceID field
+	if instanceId, ok, fieldErr := instanceID(res.Attributes()); fieldErr != nil {
```

Comment:
Is there any chance we could follow the naming conventions laid out in our style guidelines? https://datadoghq.atlassian.net/wiki/spaces/OTEL/pages/2580119947/Go+Coding+Guidelines#Naming-things

I feel like this "long variable names" style is more specific to languages like Java and less so to Go. I am aware this habit is prevalent in the OpenTelemetry repository, but I don't agree we should write less idiomatic Go code because that org. is doing so.

However, happy to debate and change the style guide at any time.

Reply:
Do you have a specific name suggestion for this one in particular? I can't think of a good short name.

Reply:
```
if iid, ok, err2 ...
```
So I guess `id`, `iid`, `v`, all work here...

Is something I saw outside in the while. `err2` being "error nested at level 2". My comment was more general for everywhere.

Reply:
Done

Reply:
Done for instance id and EC2 hostname. I think for the most part the rest have been changed on previous commits, but feel free to reopen if there are any missing